R/utilities-mass-balance.R

```diff
@@ -8,277 +8,238 @@
+  allMoleculePaths <- ospsuite::getAllMoleculePathsIn(simulation)
   ospsuite::clearOutputs(simulation)
-  for (molecule in molecules) {
-    addOutputs(quantitiesOrPaths = molecule, simulation = simulation)
-  }
-
+  addOutputs(quantitiesOrPaths = allMoleculePaths, simulation = simulation)
   simulationResults <- ospsuite::runSimulation(simulation)
-  simulationResultsOutput <- ospsuite::getOutputValues(
-    simulationResults = simulationResults,
-    quantitiesOrPaths = simulationResults$allQuantityPaths
-  )
 
-  # Create a data.frame with full string path and separates the last 3 elements of paths
-  # to perform filtering and grouping of for the final plot
-  pathsArray <- NULL
-  for (path in simulationResults$allQuantityPaths) {
-    endOfPath <- utils::tail(ospsuite::toPathArray(path), 3)
-    pathsArray <- rbind.data.frame(
-      pathsArray,
-      data.frame(
-        "parentCompartmentName" = endOfPath[1],
-        "subCompartmentName" = endOfPath[2],
-        "compoundName" = endOfPath[3],
-        "path" = path,
-        stringsAsFactors = FALSE
+  allAppliedCompoundNames <- simulation$allXenobioticFloatingMoleculeNames()
+  # Default configuration behavior integrates all xenobiotic molecules in mass balance
+  # Note that a list of lists is expected from the json file,
+  # thus the same format is used for default configuration
+  massBalanceSettings <- list(list(
+    Name = paste("Mass Balance for", paste(allAppliedCompoundNames, collapse = ", ")),
+    Molecules = allAppliedCompoundNames
+  ))
+  massBalanceSettings <- structureSet$simulationSet$massBalanceSettings %||% massBalanceSettings
+
+  massBalanceResults <- list()
+  for (plotSettings in massBalanceSettings) {
+    # Sub section using settings name
+    if(!isEmpty(plotSettings$Name)){
+      sectionId <- defaultFileNames$resultID(length(massBalanceResults) + 1, "mass_balance")
+      massBalanceResults[[sectionId]] <- saveTaskResults(
+        id = sectionId,
+        textChunk = paste("###", plotSettings$Name, anchor(sectionId)),
+        includeTextChunk = TRUE
       )
+    }
+    compoundNames <- as.character(plotSettings$Molecules)
+    massBalanceData <- getMassBalanceData(
+      groupings = plotSettings$Groupings,
+      compoundNames = compoundNames,
+      simulation = simulation,
+      simulationResults = simulationResults
     )
-  }
-
-  # In the Matlab version, there are multiple groupings performed at that point:
-  # 1) Saliva is excluded from the mass balance calculation
-  # 2) Lumen is a parent compartment that regroups many sub-compartments but Feces
-  # 3) All the rest of the sub-compartments are used normally
 
-  # 1) Exclude Saliva
-  salivaExclusionCondition <- !pathsArray[, "parentCompartmentName"] %in% "Saliva"
-  pathsArray <- pathsArray[salivaExclusionCondition, ]
+    # Get drugMass over time for compoundNames to normalize the mass balance plots
+    applications <- simulation$allApplicationsFor(head(compoundNames, 1))
```

Comment:
why `head(compoundNames, 1)` ?? It could retrieve a molecule which is not even being applied.
First we need to build an intersection between `compoundNames `and `allAppliedCompoundNames`
After that, there are 3 cases:
1. The intersection list is empty - Error
2. The intersection list contains exactly 1 molecule - take this molecule for `simulation$allApplicationsFor()`
3. The intersection list contains > 1 molecule: take the **union** of applications for all applied molecules and additionally show a warning (">1 applied molecule in the mass balance grouping, results might be innacurate")

Reply:
I agree with 1. and 2.

Regarding 3., there is currently an issue with the `{ospsuite}` package: if `compoundNames` includes a molecule and its metabolites. We get a duplication of the applications because the method retrieves the same applications for the molecule and the metabolites.

Reply:
ok, let's handle 3. as 1. for now: error
The use cases where 3. makes sense are rare anyways.

Reply:
For this reason, I just kept the first compound name thinking that the other will likely be the metabolites and get the same applications.
(However since applications are lists of R6/.NET objects, assessing if they are the same might be tricky)

Reply:
I would not rely on this and filter out only those molecules which are applied. 
And then check if the remaining list contains exactly 1 molecule or not.